File: python/2015/d14.py
import sys
import re

test = [
    ('Comet', 14, 10, 127),
    ('Dancer', 16, 11, 162),
]

r = r'(\w+) can fly (\d+) km/s for (\d+) seconds, but then must rest for (\d+) seconds.'

def load():
    with open(sys.argv[1], 'r') as f:
        lines = f.read().splitlines()
    data = []
    for line in lines:
        row = re.findall(r, line)[0]
        data.append([row[0]] + [int(x) for x in row[1:]])
    print data
    return data

def step(racers, state):
    best = 0
    for i in range(len(racers)):
        name, speed, fly_dur, rest_dur = racers[i]
        moving, distance, rest_time, fly_time, score = state[i]
        if moving:
            fly_time += 1
            distance += speed
            if fly_time >= fly_dur:
                moving = 0
                rest_time = 0
        else:
            rest_time += 1
            if rest_time >= rest_dur:
                moving = 1
                fly_time = 0

        state[i] = moving, distance, rest_time, fly_time, score
        if distance > best:
            best = distance

    for i in range(len(racers)):
        if state[i][1] == best:
            tmp = list(state[i])
            tmp[-1] += 1
            state[i] = tuple(tmp)

    return state

def race(racers, tmax):
    state = [(1, 0, 0, 0, 0)]*len(racers)
    for t in range(tmax):
        state = step(racers, state)
    return state

def winning_dist(racers, n):
    return sorted([d for (_, d, _, _, _) in race(racers, n)])[-1]

def winning_score(racers, n):
    return sorted([d for (_, _, _, _, d) in race(racers, n)])[-1]

data = load()
print 'Part 1:', winning_dist(data, 2503)
print 'Part 2:', winning_score(data, 2503)

File: src/Y2015/D14.hs
module Y2015.D14 where

import System.Environment

runLen :: Int
runLen = 2503

type Racer = (Int, Int, Int)

data State = State { moving   :: Bool,
                     distance :: Int,
                     restTime :: Int,
                     flyTime  :: Int,
                     score    :: Int
                   } deriving (Show)

initial :: State
initial = State { moving = True,
                  distance = 0,
                  restTime = 0,
                  flyTime = 0,
                  score = 0
                }


parse :: [String] -> Racer
parse [_, "can", "fly", dist, "km/s", "for", t1, "seconds,",
       "but", "then", "must", "rest", "for", t2, "seconds."] =
            (read dist, read t1, read t2)
parse _ = error "Parse error on input."


tryStop :: Racer -> State -> State
tryStop (_, flyDur, _) state
    | flyTime state >= flyDur
        = state { moving = False, restTime = 0 }
    | otherwise = state


tryStart :: Racer -> State -> State
tryStart (_, _, restDur) state
    | restTime state >= restDur
        = state { moving = True, flyTime = 0 }
    | otherwise = state


stepRacer :: Racer -> State -> State
stepRacer racer@(speed,_,_) state
    | moving state
        = tryStop racer (state { flyTime = flyTime state + 1,
                                 distance = distance state + speed })
    | otherwise = tryStart racer (state { restTime = restTime state + 1 })


updateScores :: Int -> [State] -> [State]
updateScores high = map update'
    where update' s
            | distance s == high = s { score = score s + 1 }
            | otherwise = s


step :: [Racer] -> [State] -> [State]
step racers states = updateScores best stepped
    where stepped = zipWith stepRacer racers states
          best = maximum [distance s | s <- stepped]


findSolution :: String -> (Int, Int)
findSolution str = (bestDist, bestScore)
    where racers = map (parse . words) (lines str)
          states = replicate (length racers) initial
          final = iterate (step racers) states !! runLen
          bestDist = maximum [distance s | s <- final]
          bestScore = maximum [score s | s <- final]

main :: IO ()
main = do
    [path] <- getArgs
    contents <- readFile path
    let (part1, part2) = findSolution contents
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2
